File: backend/app/schemas/mapa.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict


# ── Shared enums as literals ──────────────────────────────────────────────────

TipoPotrero = str  # "agricultura" | "ganaderia" | "mixto"
EstadoPasto = str  # "bueno" | "regular" | "malo"
EspecieAnimal = str  # "bovino" | "ovino" | "equino" | "porcino" | "otro"
TipoPunto = str  # "bebedero" | "casa" | "sombra" | "comedero"
EstadoMovimiento = str  # "programado" | "ejecutado" | "cancelado"


# ── Potrero ───────────────────────────────────────────────────────────────────

class PotreroCreate(BaseModel):
    nombre: str
    geometria: dict[str, Any]  # GeoJSON Polygon
    tipo: str
    estado_pasto: str
    hectareas: Optional[Decimal] = None
    tiene_suplementacion: bool = False
    suplementacion_detalle: Optional[str] = None
    tiene_franjas: bool = False
    cantidad_franjas: Optional[int] = None
    franjas_usadas: Optional[int] = None
    dias_por_franja: Optional[int] = None
    observaciones: Optional[str] = None
    cultivo: Optional[str] = None
    es_primera: Optional[bool] = None
    fecha_siembra: Optional[date] = None
    coneat: Optional[Decimal] = None
    kg_producidos_anio: Optional[Decimal] = None


class PotreroUpdate(BaseModel):
    nombre: Optional[str] = None
    geometria: Optional[dict[str, Any]] = None
    tipo: Optional[str] = None
    estado_pasto: Optional[str] = None
    hectareas: Optional[Decimal] = None
    tiene_suplementacion: Optional[bool] = None
    suplementacion_detalle: Optional[str] = None
    tiene_franjas: Optional[bool] = None
    cantidad_franjas: Optional[int] = None
    franjas_usadas: Optional[int] = None
    dias_por_franja: Optional[int] = None
    observaciones: Optional[str] = None
    cultivo: Optional[str] = None
    es_primera: Optional[bool] = None
    fecha_siembra: Optional[date] = None
    coneat: Optional[Decimal] = None
    kg_producidos_anio: Optional[Decimal] = None


class PotreroRead(BaseModel):
    id: int
    user_id: int
    nombre: str
    geometria: dict[str, Any]  # GeoJSON
    tipo: str
    estado_pasto: str
    hectareas: Optional[Decimal]
    tiene_suplementacion: bool
    suplementacion_detalle: Optional[str]
    tiene_franjas: bool
    cantidad_franjas: Optional[int]
    franjas_usadas: Optional[int]
    dias_por_franja: Optional[int]
    observaciones: Optional[str]
    en_descanso: bool
    fecha_descanso: Optional[date]
    cultivo: Optional[str] = None
    es_primera: Optional[bool] = None
    fecha_siembra: Optional[date] = None
    coneat: Optional[Decimal] = None
    kg_producidos_anio: Optional[Decimal] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Animal ────────────────────────────────────────────────────────────────────

class AnimalCreate(BaseModel):
    especie: str
    cantidad: int
    raza: Optional[str] = None


class AnimalUpdate(BaseModel):
    especie: Optional[str] = None
    cantidad: Optional[int] = None
    raza: Optional[str] = None


class AnimalRead(BaseModel):
    id: int
    potrero_id: int
    user_id: int
    especie: str
    cantidad: int
    raza: Optional[str]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── PuntoInteres ──────────────────────────────────────────────────────────────

class PuntoInteresCreate(BaseModel):
    nombre: str
    tipo: str
    geometria: dict[str, Any]  # GeoJSON Point
    potrero_id: Optional[int] = None


class PuntoInteresRead(BaseModel):
    id: int
    user_id: int
    potrero_id: Optional[int]
    nombre: str
    tipo: str
    geometria: dict[str, Any]  # GeoJSON Point
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── MovimientoGanado ──────────────────────────────────────────────────────────

class MovimientoCreate(BaseModel):
    potrero_origen_id: int
    potrero_destino_id: int
    cantidad: int
    especie: str
    fecha_programada: date
    ejecutar_ahora: bool = False
    notas: Optional[str] = None


class MovimientoRead(BaseModel):
    id: int
    user_id: int
    potrero_origen_id: int
    potrero_destino_id: int
    potrero_origen_nombre: str
    potrero_destino_nombre: str
    cantidad: int
    especie: str
    fecha_programada: date
    fecha_ejecutada: Optional[date]
    estado: str
    notas: Optional[str]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── AplicacionPotrero ─────────────────────────────────────────────────────────

class AplicacionCreate(BaseModel):
    producto: str
    fecha_aplicacion: date
    costo: Optional[Decimal] = None
    moneda: str = "UYU"
    observaciones: Optional[str] = None


class AplicacionRead(BaseModel):
    id: int
    potrero_id: int
    producto: str
    fecha_aplicacion: date
    costo: Optional[Decimal]
    moneda: str
    observaciones: Optional[str]
    registro_id: Optional[int]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Rentabilidad ──────────────────────────────────────────────────────────────

class RentabilidadPotrero(BaseModel):
    potrero_id: int
    nombre: str
    hectareas: Optional[Decimal]
    total_ingresos: Decimal
    total_gastos: Decimal
    balance: Decimal
    rentabilidad_pct: Optional[Decimal]
    cantidad_animales: int
    margen_bruto_ha: Optional[Decimal]
    carga_animal_ug_ha: Optional[Decimal]
    produccion_kg_ha: Optional[Decimal]
    coneat: Optional[Decimal] = None

File: frontend/src/types/mapa.ts
// ── GeoJSON primitives ────────────────────────────────────────────────────────

export interface GeoJSONPolygon {
  type: "Polygon";
  coordinates: number[][][];
}

export interface GeoJSONPoint {
  type: "Point";
  coordinates: [number, number]; // [lng, lat]
}

// ── Enums ─────────────────────────────────────────────────────────────────────

export type TipoPotrero = "agricultura" | "ganaderia" | "mixto";
export type EstadoPasto = "bueno" | "regular" | "malo";
export type EspecieAnimal = string;
export type TipoPunto = "bebedero" | "casa" | "sombra" | "comedero";
export type EstadoMovimiento = "programado" | "ejecutado" | "cancelado";

// ── Potrero ───────────────────────────────────────────────────────────────────

export interface Potrero {
  id: number;
  user_id: number;
  nombre: string;
  geometria: GeoJSONPolygon;
  tipo: TipoPotrero;
  estado_pasto: EstadoPasto;
  hectareas: number | null;
  tiene_suplementacion: boolean;
  suplementacion_detalle: string | null;
  tiene_franjas: boolean;
  cantidad_franjas: number | null;
  franjas_usadas: number | null;
  dias_por_franja: number | null;
  observaciones: string | null;
  en_descanso: boolean;
  fecha_descanso: string | null;
  cultivo?: string | null;
  es_primera?: boolean | null;
  fecha_siembra?: string | null;
  coneat?: number | null;
  kg_producidos_anio?: number | null;
  created_at: string;
}

export interface PotreroCreate {
  nombre: string;
  geometria: GeoJSONPolygon;
  tipo: TipoPotrero;
  estado_pasto: EstadoPasto;
  hectareas?: number;
  tiene_suplementacion?: boolean;
  suplementacion_detalle?: string;
  tiene_franjas?: boolean;
  cantidad_franjas?: number;
  franjas_usadas?: number;
  dias_por_franja?: number | null;
  observaciones?: string;
  cultivo?: string | null;
  es_primera?: boolean | null;
  fecha_siembra?: string | null;
  coneat?: number | null;
  kg_producidos_anio?: number | null;
}

export interface PotreroUpdate {
  nombre?: string;
  geometria?: GeoJSONPolygon;
  tipo?: TipoPotrero;
  estado_pasto?: EstadoPasto;
  hectareas?: number | null;
  tiene_suplementacion?: boolean;
  suplementacion_detalle?: string | null;
  tiene_franjas?: boolean;
  cantidad_franjas?: number | null;
  franjas_usadas?: number | null;
  dias_por_franja?: number | null;
  observaciones?: string | null;
  cultivo?: string | null;
  es_primera?: boolean | null;
  fecha_siembra?: string | null;
  coneat?: number | null;
  kg_producidos_anio?: number | null;
}

export interface AplicacionPotrero {
  id: number;
  potrero_id: number;
  producto: string;
  fecha_aplicacion: string;
  costo: number | null;
  moneda: string;
  observaciones: string | null;
  registro_id: number | null;
  created_at: string;
}

export interface AplicacionCreate {
  producto: string;
  fecha_aplicacion: string;
  costo?: number | null;
  moneda?: string;
  observaciones?: string;
}

// ── Animal ────────────────────────────────────────────────────────────────────

export interface Animal {
  id: number;
  potrero_id: number;
  user_id: number;
  especie: EspecieAnimal;
  cantidad: number;
  raza: string | null;
  created_at: string;
}

export interface AnimalCreate {
  especie: EspecieAnimal;
  cantidad: number;
  raza?: string;
}

export interface AnimalUpdate {
  especie?: EspecieAnimal;
  cantidad?: number;
  raza?: string | null;
}

// ── PuntoInteres ──────────────────────────────────────────────────────────────

export interface PuntoInteres {
  id: number;
  user_id: number;
  potrero_id: number | null;
  nombre: string;
  tipo: TipoPunto;
  geometria: GeoJSONPoint;
  created_at: string;
}

export interface PuntoInteresCreate {
  nombre: string;
  tipo: TipoPunto;
  geometria: GeoJSONPoint;
  potrero_id?: number;
}

// ── MovimientoGanado ──────────────────────────────────────────────────────────

export interface MovimientoGanado {
  id: number;
  user_id: number;
  potrero_origen_id: number;
  potrero_destino_id: number;
  potrero_origen_nombre: string;
  potrero_destino_nombre: string;
  cantidad: number;
  especie: EspecieAnimal;
  fecha_programada: string; // "YYYY-MM-DD"
  fecha_ejecutada: string | null;
  estado: EstadoMovimiento;
  notas: string | null;
  created_at: string;
}

export interface MovimientoCreate {
  potrero_origen_id: number;
  potrero_destino_id: number;
  cantidad: number;
  especie: EspecieAnimal;
  fecha_programada: string;
  ejecutar_ahora?: boolean;
  notas?: string;
}
